[WIP/TEST] Mdd rds split testing wip  - demo branch for future RDS

mikedlr opened the pull request:
##### SUMMARY

Due to the requirement for a PR to only contain a single module, each of the new modules for RDS is in a separate PR, however they need to be tested together.  This is a branch which has all those PRs merged (yes - git merged - this branch should ever go back to devel) into one.  

##### ISSUE TYPE

 - Feature Pull Request
 - New Module Pull Request

##### COMPONENT NAME

rds various

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (mdd_rds_split_testing_wip bde482f5bf) last updated 2017/09/22 11:43:04 (GMT +100)
  config file = None
  configured module search path = [u'/home/mikedlr/dev/ansible/ansible-dev/library']
  ansible python module location = /home/mikedlr/dev/ansible/ansible-dev/lib/ansible
  executable location = /home/mikedlr/dev/ansible/ansible-dev/bin/ansible
  python version = 2.7.5 (default, Nov  6 2016, 00:28:07) [GCC 4.8.5 20150623 (Red Hat 4.8.5-11)]
```

##### ADDITIONAL INFORMATION

The integration tests on this branch may run for an impractically long time, up to 2hours, especially as various new cases are being added. 

*DO NOT MERGE THIS TO DEVEL*

willthames committed:
Add test suite for RDS modules

Commit:
rds_instance: storage_type and tags

Support storage_type option

modify_db_instance doesn't allow tag manipulation
so we have to do it through the RDS tag APIs instead.

Commit:
rds_instance: fix MultiAZ capitilization

Allow MultiAZ manipulation

Commit:
rds_instance: improve compliance with standards

Update imports and documentation to better meet
Ansible standards

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

willthames committed:
RDS: split into more sensible modules

RDS instance and snapshot facts modules
Move common code into module_utils/rds.py

This work is a precursor to new rds_snapshot
and rds_instance modules

mikedlr committed:
Start writing tests for rds module

Commit:
Provide all returned detail as facts

Commit:
Provide all returned detail as facts

Commit:
Provide all returned detail as facts

Commit:
Provide all returned detail as facts

Commit:
Provide all returned detail as facts

Commit:
rds facts: Update to latest metadata standards

Commit:
rds facts: Update to latest metadata standards

mikedlr committed:
rds_* module integration tests - a set of tests to check compatibility between new and old

Commit:
Add rds_snapshot module

mikedlr committed:
rds_instance module - fix various parameter handling cases from trying to run integration test cases

mikedlr committed:
rds_instance module - fix various parameter handling cases from trying to run integration test cases

mikedlr committed:
rds_instance module - fix various parameter handling cases from trying to run integration test cases

mikedlr committed:
rds_instance module - fix various parameter handling cases from trying to run integration test cases

mikedlr committed:
rds_instance module - fix various parameter handling cases from trying to run integration test cases

Commit:
Update rds facts tests to use new modules

mikedlr committed:
rds_instance module, rds in mod utils and int tests - small fixes to make tests go

mikedlr committed:
rds_instance module, rds in mod utils and int tests - small fixes to make tests go

mikedlr committed:
rds_instance module - refresh instance in case of state just before waiting - seems to fix test

mikedlr committed:
Add some simple unit tests of the diff functionality - failing at this time to show problems

mikedlr committed:
rds_instance and related modules - Varied fixes to get more integration tests working

mikedlr committed:
rds_instance and related modules - Varied fixes to get more integration tests working

mikedlr committed:
rds_instance - fixes & tests to ensure wait instance and changed state work

mikedlr committed:
rds_instance - fixes & tests to ensure wait instance and changed state work

mikedlr committed:
rds_instance - fixes & tests to ensure wait instance and changed state work

mikedlr committed:
rds_instance - fixes & tests to ensure wait instance and changed state work

mikedlr committed:
rds_instance - fixes & tests to ensure wait instance and changed state work

mikedlr committed:
rds_instance - port parameter is optional so ignore if missing

mikedlr committed:
rds_instance integration tests - fix non modify and disable replica promition

mikedlr committed:
rds_instance module - full traceback of exceptions in fail_json better error reporting

mikedlr committed:
rds_instance - add import for traceback missing from previous commit

mikedlr committed:
rds_instance module - internally consistent error messages via fail_json in case of AWS exceptions

mikedlr committed:
partial PEP8 cleanup - flake8 is not reporting errors for some reason

Commit:
rds_snapshot_facts - fix pep8 issues

Commit:
Tidy up flake8 violations in module_utils.rds

Commit:
Tidy up flake8 violations in module_utils.rds

Commit:
Tidy up flake8 violations in module_utils.rds

Commit:
Tidy up flake8 violations in module_utils.rds

mikedlr committed:
rds_instance - test and parameter fixes required to get rds_old_new compatibility tests to pass

mikedlr committed:
rds_instance - test and parameter fixes required to get rds_old_new compatibility tests to pass

mikedlr committed:
rds_instance integration test - increase wait timeout for snapshot restoration

mikedlr committed:
rds_instance module - make security group parameters work

mikedlr committed:
rds_instance module - improve error messages for easier debugging

mikedlr committed:
rds_instance module - allow publicly_accessible parameter during modify

mikedlr committed:
rds_instance module - flake8 and autopep8 cleanup - strongly

mikedlr committed:
rds_instance module - fix wait fail error message to show actual instance state

mikedlr committed:
rds_instance module - turn on compatibility tests - n.b. long

mikedlr committed:
rds_instance module - bring in review comments from ryansb

mikedlr committed:
move all rds integration tests from roles to target directories

mikedlr committed:
rds_instance integration tests - update for new integration testing framework

mikedlr committed:
rds integration tests - delete rds roles from old test playbook

mikedlr committed:
rds modules and tests - pep8 / flake8 error fixes

mikedlr committed:
rds modules and tests - pep8 / flake8 error fixes

mikedlr committed:
rds modules and tests - pep8 / flake8 error fixes

mikedlr committed:
rds modules and tests - pep8 / flake8 error fixes

mikedlr committed:
rds modules and tests - further pep8 and sanity fixes including docs

mikedlr committed:
rds modules and tests - further pep8 and sanity fixes including docs

mikedlr committed:
rds modules and tests - further pep8 and sanity fixes including docs

mikedlr committed:
rds modules and tests - further pep8 and sanity fixes including docs

mikedlr committed:
rds modules and tests - further pep8 and sanity fixes including docs

mikedlr committed:
rds instance - validation fixes with nasty hack for from_future

mikedlr committed:
rds instance - validation fixes with nasty hack for from_future

mikedlr committed:
rds instance - continued cleanup to appease the CI gods

mikedlr committed:
rds instance - continued cleanup to appease the CI gods

mikedlr committed:
first attempt at fixing ordering in Python3 mocks

mikedlr committed:
rds instance - continued cleanup to appease the CI gods

mikedlr committed:
rds instance - continued cleanup to appease the CI gods

mikedlr committed:
rds_instance - pycodestyle fixes and python3 (mock) compatibility fixes

mikedlr committed:
rds_instance - add more operator functions to to try to get orderable types

mikedlr committed:
rds instance - continued cleanup to appease the CI gods

mikedlr committed:
module utils - move rds.py into recently agreed aws subdirectory

mikedlr committed:
module utils - move rds.py into recently agreed aws subdirectory

mikedlr committed:
module utils - move rds.py into recently agreed aws subdirectory

mikedlr committed:
module utils - move rds.py into recently agreed aws subdirectory

mikedlr committed:
rds_instance_facts module - move rds import into recently agreed aws subdirectory

mikedlr committed:
rds_snapshot_facts module - move rds import into recently agreed aws subdirectory

mikedlr committed:
rds_instance module - move rds import into recently agreed aws subdirectory

mikedlr committed:
rds_snapshot module - move rds import into recently agreed aws subdirectory

mikedlr committed:
rds_instance tests - move rds import into recently agreed aws subdirectory

mikedlr committed:
rds integration tests - Remove accidentally included README file

mikedlr committed:
rds snapshot facts - document current lack of support for Aurora

mikedlr committed:
rds utils tests - first version as a design document

mikedlr committed:
rds utils - rewriting for new functions to handle rds data better

mikedlr committed:
rds snapshot module - fixes responding to shertel review including move to new AnsibleAWSModule

mikedlr committed:
rds instance facts module - part fixes responding to shertel review including move to new AnsibleAWSModule

mikedlr committed:
rds utils - clean out remaining old object definitions

mikedlr committed:
rds utils - get through to having working unit tests

Commit:
Simplify rds common module and add tests

Rely on boto3 structure for RDS snapshot and remove RDSSnapshot class.
Add tests for RDS snapshot
Ensure IAM configuration is correct for tests

mikedlr committed:
rds utils - aggressive fixing of pep8 and other issues - unit tests work on python3 as well

mikedlr committed:
Revert "Simplify rds common module and add tests"

This reverts commit b268b34ced66e941abf1b47f402a60c3ef0b3d1d.

For now - to be discussed with Will.
- Integration tests already exist in separate PR and do more
- Simplification is also being done in separate
- Plan to have separate functions for converting instance to facts

mikedlr committed:
from willthames - Ensure IAM configuration is correct for tests

mikedlr committed:
rds_instance_facts module - minor changes based on willthames comments - more needed later

mikedlr committed:
rds utils - add various constant data lists - rename functions and various minor fixes

mikedlr committed:
rds_instance - refactoring towards match with aws/rds utils update

mikedlr committed:
rds_instance - refactoring towards match with aws/rds utils update

mikedlr committed:
rds_instance - select_parameters() for API call prep + get status from dict not object

mikedlr committed:
rds_instance - major refactor with some rewrite - move to more standard parameter naming

mikedlr committed:
rds integration tests - update arguments for new rds_instance

mikedlr committed:
rds_instance - unpack return dict calling exit_json

mikedlr committed:
rds_snapshot module - begin refactoring to work with new rds utils

mikedlr committed:
rds integration tests - update arguments for new rds_snapshot

mikedlr committed:
rds_snapshot module - more updates for new rds utils - fix delete

mikedlr committed:
rds_instance - build tags functionality with unit test

mikedlr committed:
rds_instance - add handling of empty (cleanout) and None (leave alone) tags values

mikedlr committed:
rds_instance - conversion to AnsibleAWSModule

mikedlr committed:
rds_instance - new argument handling - failing unit test

mikedlr committed:
rds_instance - new argument handling - failing unit test

mikedlr committed:
rds modutils - fixes according to review from shertel

mikedlr committed:
rds_instance module - review from shertel + refactoring continued till unit tests pass

mikedlr committed:
rds_instance module - remove left over pdb in tests and allow differnt key order for python3

mikedlr committed:
rds_instance tests pep8 fixes

mikedlr committed:
rds_instance - fix all parameter names to match AWS standard & improve case convert

mikedlr committed:
rds integration tests - further update of arguments for new rds_snapshot

mikedlr committed:
rds_instance_facts - unit test and fixes arising from making it work

mikedlr committed:
rds_snapshot - unit test improvements and fixes arising from making them pass

mikedlr committed:
rds_snapshot_facts - add simple unit test and fixes arising

mikedlr committed:
rds_snapshot_facts - bug fix for multiple snapshot handling

mikedlr committed:
rds_instance - improve handling of snapshots

mikedlr committed:
rds_instance integration tests - fix naming parameter to match module

mikedlr committed:
rds_instance - fix modify in case where no rename is taking place

mikedlr committed:
rds_instance - fix modify in case where no rename is taking place

mikedlr committed:
rds_instance and tests - wrong database state forces retry

mikedlr committed:
rds_instance and tests - many more fixes from integration tests and unit tests to match

mikedlr committed:
rds_instance and tests - fix source_instance to correct name & handling of missing immutable variables

mikedlr committed:
rds_instance_facts - fix metadata_version

mikedlr committed:
rds_snapshot - will be added in version 2.5 not 2.4

mikedlr committed:
rds_snapshot_facts - will be added in version 2.5 not 2.4

mikedlr committed:
rds integration tests - give full data for replica creation

mikedlr committed:
rds_instance and tests - fixes to make integration tests pass including elimination of many arg lists

mikedlr committed:
rds_instance and tests - fixes to match static tests

mikedlr committed:
rds_instance and tests - further fixes to fix CI complaints

mikedlr committed:
rds integration tests - add test for promotion of RDS replica

mikedlr committed:
rds_instance module - add check mode - with tests

mikedlr committed:
rds split integration tests -  imporoved verification of check mode

mikedlr committed:
rds_instance_facts module and tests - return only exact instace asked for, not all matches

mikedlr committed:
rds module utils - eliminate unused PARAMETER_MAP variable

mikedlr committed:
rds_instance_facts module and tests - fixup: return only exact instance asked for, not all matches

mikedlr committed:
rds_instance - elimination of remaining static data - create uses botocore data

mikedlr committed:
rds_instance - further required argument fixing

mikedlr committed:
rds_instance - fix handling of instance state comparsion by geting metadata from botocore

mikedlr committed:
rds_instance - fix handling of immutable keys and removal of tags from modify calls

mikedlr committed:
rds_instance - pep8 fixes

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1178 @@
+       done until long after the module is run (e.g. at the next maintainance window).  If
+       you want to have an immediate change and see the results then you should give both
+       the 'wait' and the 'apply_immediately' parameters.  In this case, when the module
+       returns
```
This is an unfinished sentence

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1178 @@
+       happens and should reflect the status after renaming is applied but the instance
+       state is likely to continue to change afterwards.  Please do not rely on the return
+       value to match the status soon afterwards.
+     - In the case that apply_immediately is not given then the return value from
```
Partial sentence

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1178 @@
+                sqlserver-se', 'sqlserver-ex', 'sqlserver-web', 'postgres', 'aurora']
+  allocated_storage:
+    description:
+      - Size in gigabytes of the initial storage for the DB instance.  See
```
Double space before 'See'

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1178 @@
+
+    instance_id = module.params.get('db_instance_identifier')
+    params['DBInstanceIdentifier'] = instance_id
+
```
This method need:

```
params['Tags'] = ansible_dict_to_boto3_tag_list(params['Tags'])
```

and some associated tagging tests :)

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1178 @@
+
+    # modify_db_instance does not cope with DBSubnetGroup not moving VPC!
+    try:
+        if (before_facts['db_subnet_group_name'] == params.get('DBSubnetGroupName')):
```
This should be:
```
before_facts['db_subnet_group']['db_subnet_group_name']
```

Is this missing a test?

willthames commented:
rds_instance idempotency checks are failing - seeing `changed: true` when repeating the same instance

willthames committed:
Revert iam_managed_policy to upstream version

willthames committed:
More IAM policies for RDS testing

willthames committed:
Improve rds_instance tag handling

Ensure tags are converted when creating an instasnce
Allow purge_tags parameter
Return tags in results

willthames committed:
Improve rds_instance modifications

Prevent more cases of modify_db_instance where no modification is required

willthames committed:
Improve rds_instance test coverage

* Add more tag testing
* Add rds_subnet_group and rds_param_group idempotency testing

mikedlr committed:
rds_instance - clean up asserts in unit tests avoiding newer constructions, clean documentation

mikedlr committed:
rds_instance - huge amount of cleanups to get tests passing after recent changes + revert to working camel case conversions

mikedlr commented:
@willthames I have merged in the pull request you did on my other account - this was quite a bit of work to get the unit tests back working, but they found quite a few problems in both your code and mine but in the end I think it sums up to quite a bit of improvement.  

Don't worry about where you put PRs; I can move them around as needed, however please do make sure you let me know on IRC since more likely than not I will miss an email.  

I had to revert your changes to the camelisation back to using my version since it wasn't handling DBInstanceIdentifier correctly (converting to DbInstanceIdentifier).

I never reproduced the lack of idempotency.  I guess it's now fixed with the merge of your code.  Could you please confirm if you see it again?

mikedlr committed:
ec2.py - fix dangerous default value

mikedlr committed:
ec2.py - fix dangerous default value - cuter fix to get CI to rerun

mikedlr commented:
rerunning CI (again again)

ryansb commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,168 @@
+    if not instance:
+        return instance
+    assert 'DBInstanceIdentifier' in instance, "instance argument was not a valid instance"
+    d = camel_dict_to_snake_dict(instance, ignore_list=['Tags'])
```
Do this as a oneline instead: `return camel_dict_to_snake.....` no need to assign.

ryansb commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,168 @@
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'DBSnapshotNotFound':
+            return None
+        else:
```
Since you return above, unnest the `raise` here.

ryansb commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,168 @@
+
+def snapshot_to_facts(snapshot):
+    assert 'DBSnapshotIdentifier' in snapshot, "snapshot argument was not a valid snapshot"
+    d = camel_dict_to_snake_dict(snapshot)
```
Same: oneline this since the assignment is unneeded.

ryansb commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,168 @@
+    for k in compare_keys:
+        if state_a.get(k) != state_b.get(k):
+            if state_b.get(k) is None and k not in remove_if_null:
+                pass
```
This can instead be `continue` and let the `else`  unnest.

ryansb commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,168 @@
+
+    state_a['db_parameter_group_name'] = state_a.get('db_parameter_group', {}).get('db_parameter_group_name')
+    if state_a['db_parameter_group_name']:
+        del(state_a['db_parameter_group'])
```
`del` is a builtin, not a function. These parens don't have to be here

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance_facts.py:
```diff
@@ -0,0 +1,150 @@
+
+    marker = ''
+    try:
+        results = conn.describe_db_instances(Marker=marker, **params)['DBInstances']
```
We should really be using paginators where possible http://boto3.readthedocs.io/en/latest/reference/services/rds.html#paginators , especially since the users of _facts could plausibly want every RDS instance in their account. 

Also, for paginators I've grown to like using the `.search` like [this](https://github.com/ansible/ansible/pull/31997/files#diff-fc83b7d0a8f033b0b1b816e05199c5a2R255). It lets you avoid putting pages back together later, so you could here use: 

```
instances = conn.get_paginator('describe_db_instances').paginate(
    **params).search("DBInstances[]")
facts =[instance_to_facts(instance) for instance in instances]
```

ryansb commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+        # Temporary until all the rds2 commands have their responses parsed
+        if resource is None:
+            module.fail_json(msg="Failed waiting for RDS snapshot %s" % resource_name)
+        resource = get_snapshot(conn, resource_name)
```
If resource doesn't exist, this will return `None` and result in a keyerror up top of the loop. Can you add a guard here where it's transformed to an empty dict, and use `.get()` in the loop control?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+    try:
+        await_resource(conn, result, 'deleted', module)
+        return dict(changed=True)
+    except Exception as e:
```
We can't catch arbitrary `Exception` here, then rely on `.response` existing. Can you change this to the botocore exceptions you mean to be catching?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_snapshot_facts.py:
```diff
@@ -0,0 +1,160 @@
+    results = list()
+    while True:
+        try:
+            response = conn.describe_db_snapshots(Marker=marker, **params)
```
Same here - pagination is the greatest http://boto3.readthedocs.io/en/latest/reference/services/rds.html#paginators

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+DOCUMENTATION = '''
+---
+module: rds_instance
+short_description: create, delete, or modify an Amaz`on rds instance
```
Extra "`" in "Amazon" here

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+short_description: create, delete, or modify an Amaz`on rds instance
+description:
+     - Creates, deletes, or modifies rds instances. When creating an instance
+       it can be either a new instance or a read-only replica of an exising instance.
```
"Creates, deletes, or modifies rds instances. This module can operate on new DB instances or on read-only replicas."

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+       done until long after the module is run (e.g. at the next maintainance window).  If
+       you want to have an immediate change and see the results then you should give both
+       the 'wait' and the 'apply_immediately' parameters.  In this case, when the module
+       returns
```
Needs to be a complete sentence. "With both 'wait' and 'apply_immediately', changes will be reflected when the module returns."

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+       you want to have an immediate change and see the results then you should give both
+       the 'wait' and the 'apply_immediately' parameters.  In this case, when the module
+       returns
+     - The behavior in the case where only one of apply_immediately or wait is given is
```
Thank you for calling this out explicitly. :+1:

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+  allocated_storage:
+    description:
+      - Size in gigabytes of the initial storage for the DB instance.  See
+        [API documentation](https://botocore.readthedocs.io/en/latest/reference/services/rds.html#RDS.Client.create_db_instance)
```
Docs don't accept markdown links - needs to be RST format.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+  master_username:
+    description:
+      - Master database username.
+  master_user_password:
```
What happens if this+the user are not specified?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+  backup_retention_period:
+    description:
+       - "Number of days backups are retained. Set to 0 to disable backups. Default is 1 day. "
+       - "Valid range: 0-35."
```
Also worth noting the downtime incurred switching the retention period from 0 to non-zero, or non-zero to 0. See docs [here](http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_WorkingWithAutomatedBackups.html#USER_WorkingWithAutomatedBackups.BackupRetention)

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    description:
+      - Name of snapshot to take when state=absent - if no snapshot name is provided then no
+        snapshot is taken.
+  db_snapshot_identifier:
```
Wouldn't this be `source_db_snapshot` since you're restoring the instance *from* that source?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+  snapshot:
+    description:
+      - snapshot provides a default for either final_db_snapshot_identifier or db_snapshot_identifier
+        allowing the same parameter to be used for both backup and restore.
```
I've never tried this - what happens when you have a `final_db_snapshot_identifier` where the snapshot exists already? This seems like it overwrites whatever's there, is that right?

s-hertel replied:
Tested this, and it results in failure:
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_pS34v7/ansible_module_rds_instance.py", line 881, in ensure_absent_db_instance
    response = conn.delete_db_instance(**params)
  File "/Users/shertel/Workspace/ansible/venv/python2.7/lib/python2.7/site-packages/botocore/client.py", line 312, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/shertel/Workspace/ansible/venv/python2.7/lib/python2.7/site-packages/botocore/client.py", line 605, in _make_api_call
    raise error_class(parsed_response, operation_name)
DBSnapshotAlreadyExistsFault: An error occurred (DBSnapshotAlreadyExists) when calling the DeleteDBInstance operation: Cannot create the snapshot because a snapshot with the identifier final-snapshot-works already exists.

fatal: [localhost]: FAILED! => {
    "changed": false,
    "error": {
        "code": "DBSnapshotAlreadyExists",
        "message": "Cannot create the snapshot because a snapshot with the identifier final-snapshot-works already exists.",
        "type": "Sender"
    },
    "invocation": {
```

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+# setting.  We use the full parameter names matching the ones AWS uses.
+- rds_instance:
+    db_instance_identifier: new-database-replica
+    wait: yes
```
Wouldn't this also require an `apply_immediately`?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+# we can always point to a more sophisticated logger when one exists.
+
+
+class the_logger():
```
This naming scheme doesn't follow convention.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+# me to fully understand but including the need to properly handle
+# security (see proposals, many tickets and a number of PRs), so we
+# make a little improvised logger here which goes through module.log
+# we can always point to a more sophisticated logger when one exists.
```
Is the purpose of this logger just to shim for `module.log` and ship to `q` when that's installed?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+#
+
+
+# Q is a simple logging framework NOT suitable for production use but handy in development.
```
Can we avoid reinventing logging, and stick to module.log? I really don't want yet-another-way-to-log in the codebase.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+        rdat = resource["PendingModifiedValues"]
+    # resource will be none if it has actually been removed - e.g. we were waiting for deleted
+    # status; maybe that should be an error in other situations?
+    if wait_timeout <= time.time() and resource is not None and resource['DBInstanceStatus'] != status:
```
Instead of checking times at the `while` control statement *and* here, just move this into the loop and have it fail from there.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+            response = conn.create_db_instance_read_replica(**params)
+            instance = get_db_instance(conn, instance_id)
+            changed = True
+        except Exception as e:
```
:-1: on catching raw `Exception` here.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    purge_tags = module.params.get('purge_tags')
+    changed = False
+
+    tags_need_modify, tags_to_delete = compare_aws_tags(current_tags, tags, purge_tags)
```
For consistence, can you name this `tags_to_modify`?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+        try:
+            conn.remove_tags_from_resource(ResourceName=db_instance_arn, TagKeys=tags_to_delete)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
```
Is there a reason not to use `fail_json_aws` here?

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    # modify_db_instance does not cope with DBSubnetGroup not moving VPC!
+    try:
+        if before_facts['db_subnet_group_name'] == params.get('DBSubnetGroupName'):
+            del(params['DBSubnetGroupName'])
```
`del` is a builtin not a function, spurious `()`

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    except KeyError:
+        pass
+    if not force_password_update:
+        try:
```
For optional-deletes, I like `params.pop('ParamNameWhatever', None` instead of catching `KeyError`

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+def ensure_absent_db_instance(module, conn):
+    main_logger.log(30, "delete_db_instance called")
+    try:
+        del(module.params['storage_type'])
```
Optional `del` could be `pop`

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'DBInstanceNotFound':
+            return dict(changed=True)
+        else:
```
Unnest `else` after `return`

s-hertel commented on test/integration/targets/rds_compat/tasks/main.yml:
```diff
@@ -0,0 +1,229 @@
+        security_token: '{{security_token}}'
+      register: rds_result
+
+    - name: When I run the new module with unchanged parameters
```
```
TASK [rds_compat : When I run the new module with unchanged parameters] **************************************************
task path: /Users/shertel/Workspace/ansible/test/integration/targets/rds_compat/tasks/main.yml:21
Using module file /Users/shertel/Workspace/ansible/lib/ansible/modules/cloud/amazon/rds_instance.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: shertel
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/shertel/.ansible/tmp/ansible-tmp-1509122757.81-106199026384651 `" && echo ansible-tmp-1509122757.81-106199026384651="` echo /Users/shertel/.ansible/tmp/ansible-tmp-1509122757.81-106199026384651 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpixDXJr TO /Users/shertel/.ansible/tmp/ansible-tmp-1509122757.81-106199026384651/rds_instance.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/shertel/.ansible/tmp/ansible-tmp-1509122757.81-106199026384651/ /Users/shertel/.ansible/tmp/ansible-tmp-1509122757.81-106199026384651/rds_instance.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-mvkyhg/coverage/injector.py /Users/shertel/.ansible/tmp/ansible-tmp-1509122757.81-106199026384651/rds_instance.py; rm -rf "/Users/shertel/.ansible/tmp/ansible-tmp-1509122757.81-106199026384651/" > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_mps9cy/ansible_modlib.zip/ansible/module_utils/aws/core.py", line 80, in __init__
    local_settings[key] = kwargs.pop(key)

fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "allocated_storage": 10,
            "aws_access_key": "ASIAISPCNEHIPUU22QBA",
            "aws_secret_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "db_engine": "postgres",
            "instance_name": "ansible-test-shertel-OSX-2-13217440-rds",
            "instance_type": "db.t2.micro",
            "password": "9UkByXOyboNnFnu1KaJX",
            "publicly_accessible": false,
            "region": "us-east-1",
            "security_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "size": 10,
            "state": "present",
            "username": "db_user",
            "wait": true,
            "wait_timeout": 1200
        }
    },
    "msg": "Unsupported parameters for (rds_instance) module: db_engine,instance_name,instance_type,password,username Supported parameters include: allocated_storage,apply_immediately,auto_minor_version_upgrade,availability_zone,aws_access_key,aws_secret_key,backup_retention_period,character_set_name,db_instance_class,db_instance_identifier,db_name,db_parameter_group_name,db_security_groups,db_snapshot_identifier,db_subnet_group_name,ec2_url,engine,engine_version,final_db_snapshot_identifier,force_failover,force_password_update,iops,license_model,log_level,master_user_password,master_username,multi_az,old_db_instance_identifier,option_group_name,port,preferred_backup_window,preferred_maintenance_window,profile,publicly_accessible,purge_tags,region,security_token,skip_final_snapshot,snapshot,source_db_instance_identifier,state,storage_type,tags,validate_certs,vpc_security_group_ids,wait,wait_timeout"
}
```

s-hertel commented on test/integration/targets/rds_instance/tasks/main.yml:
```diff
@@ -0,0 +1,539 @@
+      debug:
+        var: rds_instance
+
+    - name: get instance facts
```
```
TASK [rds_instance : get instance facts] *********************************************************************************
task path: /Users/shertel/Workspace/ansible/test/integration/targets/rds_instance/tasks/main.yml:159
Using module file /Users/shertel/Workspace/ansible/lib/ansible/modules/cloud/amazon/rds_instance_facts.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: shertel
<127.0.0.1> EXEC /bin/sh -c 'echo ~ && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/shertel/.ansible/tmp/ansible-tmp-1509123327.61-150214598672521 `" && echo ansible-tmp-1509123327.61-150214598672521="` echo /Users/shertel/.ansible/tmp/ansible-tmp-1509123327.61-150214598672521 `" ) && sleep 0'
<127.0.0.1> PUT /tmp/tmpQ0Ii77 TO /Users/shertel/.ansible/tmp/ansible-tmp-1509123327.61-150214598672521/rds_instance_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/shertel/.ansible/tmp/ansible-tmp-1509123327.61-150214598672521/ /Users/shertel/.ansible/tmp/ansible-tmp-1509123327.61-150214598672521/rds_instance_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-OM2ywK/coverage/injector.py /Users/shertel/.ansible/tmp/ansible-tmp-1509123327.61-150214598672521/rds_instance_facts.py; rm -rf "/Users/shertel/.ansible/tmp/ansible-tmp-1509123327.61-150214598672521/" > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_iYdyVC/ansible_module_rds_instance_facts.py", line 150, in <module>
    main()
  File "/tmp/ansible_iYdyVC/ansible_module_rds_instance_facts.py", line 146, in main
    module.exit_json(**instance_facts(module, conn))
  File "/tmp/ansible_iYdyVC/ansible_module_rds_instance_facts.py", line 109, in instance_facts
    assert len(results) == 1
AssertionError

fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_iYdyVC/ansible_module_rds_instance_facts.py\", line 150, in <module>\n    main()\n  File \"/tmp/ansible_iYdyVC/ansible_module_rds_instance_facts.py\", line 146, in main\n    module.exit_json(**instance_facts(module, conn))\n  File \"/tmp/ansible_iYdyVC/ansible_module_rds_instance_facts.py\", line 109, in instance_facts\n    assert len(results) == 1\nAssertionError\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
```

s-hertel replied:
If I remove the assert in the module and rerun the tests then TASK [rds_instance : check facts match expected instance] fails (provided the output of that below).

s-hertel commented on test/integration/targets/rds_instance/tasks/main.yml:
```diff
@@ -0,0 +1,539 @@
+        security_token: '{{security_token}}'
+      register: rds_instance_facts
+
+    - name: check facts match expected instance
```
```
TASK [rds_instance : check facts match expected instance] ****************************************************************
task path: /Users/shertel/Workspace/ansible/test/integration/targets/rds_instance/tasks/main.yml:168
fatal: [localhost]: FAILED! => {
    "msg": "The conditional check 'rds_instance_facts.db_instances[0].db_instance_class == \"db.t2.micro\"' failed. The error was: error while evaluating conditional (rds_instance_facts.db_instances[0].db_instance_class == \"db.t2.micro\"): list object has no element 0"
}
```

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    description:
+      - Specifies the storage type to be associated with the DB instance. C(iops) must
+        be specified if C(io1) is chosen.
+    choices: ['standard', 'gp2', 'io1' ]
```
Aurora uses the AWS proprietary storage system as opposed to EBS, so these choices don't work for that engine.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+        if module.check_mode:
+            module.exit_json(changed=True, create_db_instance_params=params)
+        try:
+            response = conn.create_db_instance(**params)
```
Standalone DB instances are not supported for aurora, so if this module should be able to create Aurora instances this needs another boto call.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    for immutable_key in rds_immutable_params:
+        if immutable_key in module.params and module.params[immutable_key] is not None:
+            before = before_facts.get(immutable_key)
+            if (module.params[immutable_key] != before):
```
I feel like there might be a number of corner cases for this comparison. The only one I found yet is for the engine: 'MySQL' (the var `before` is 'mysql' so the casing is different), but I haven't tested this thoroughly.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+        # it changes at a random time in future.
+        if update_rds_tags(module, conn, instance):
+            changed = True
+        return_dict = modify_db_instance(module, conn, instance)
```
I can't run a task like this twice:
```
    - name: ensure MySQL rds instance exists
      rds_instance:
        profile: shertel
        region: us-east-1
        state: present
        db_instance_identifier: "{{ id }}"
        db_instance_class: db.m1.small
        engine: MySQL
        allocated_storage: 10
        master_username: "{{ username }}"
        master_user_password: "{{ password }}"
        tags:
          one: two
          Three: Four
```
Running that once creates the instance. Rerunning (with commenting out the check that fails [here](https://github.com/ansible/ansible/pull/30746/files#diff-5bee6361468911fe9715542b1b45f4fbR560)) fails because Engine, required for creation isn't a valid parameter for boto's modify_db_instance().

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    ))
+
+    call_params = prepare_params_for_modify(module, conn, instance)
+    if not call_params:
```
The instance id is in call_params (and ApplyImmediately may be too) so it always shows changed=True even if there are no modifications. If there are actually no modifications this happens:
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_C0GmmZ/ansible_module_rds_instance.py", line 720, in modify_db_instance
    response = modify_the_instance(**call_params)
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_C0GmmZ/ansible_modlib.zip/ansible/module_utils/cloud.py", line 150, in retry_func
    raise e
ClientError: An error occurred (InvalidParameterCombination) when calling the ModifyDBInstance operation: No modifications were requested

fatal: [localhost]: FAILED! => {
    "changed": false,
    "error": {
        "code": "InvalidParameterCombination",
        "message": "No modifications were requested",
        "type": "Sender"
    },
...
    "msg": "trying to modify RDS instance: An error occurred (InvalidParameterCombination) when calling the ModifyDBInstance operation: No modifications were requested",
    "response_metadata": {
        "http_headers": {
            "connection": "close",
            "content-length": "289",
            "content-type": "text/xml",
            "date": "Fri, 27 Oct 2017 18:56:41 GMT",
            "x-amzn-requestid": "92043ccc-bb48-11e7-8942-3fc4dbc4cdd3"
        },
        "http_status_code": 400,
        "request_id": "92043ccc-bb48-11e7-8942-3fc4dbc4cdd3",
        "retry_attempts": 0
    }
}
```

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    params = {}
+
+    operations_model = conn._service_model.operation_model(operation)
+    relevant_parameters = operations_model.input_shape.members.keys()
```
This is nice! It looks like modify_db_instance() is missing `select_parameters_meta(module, conn, 'ModifyDBInstance')`?

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+      Application: cms
+
+# Create a read-only replica and wait for it to become available
+- rds_instance:
```
engine and db_instance_class are required right now for state: present (even if providing source_db_instance_identifier), so those could be added to these examples or changed.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+    if instance is None:
+        return_dict = create_db_instance(module, conn)
+    else:
+        if instance.get('replication_source') and not module.params.get('source_db_instance_identifier'):
```
I think 'replication_source' needs to be 'ReadReplicaDBInstanceIdentifiers'.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+  source_db_instance_identifier:
+    description:
+      - Name of the database when sourcing from a replica
+  replica:
```
This doesn't get used anywhere.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1191 @@
+  multi_az:
+    description:
+      - Specifies if this is a Multi-availability-zone deployment. Can not be used in conjunction with zone parameter.
+    choices: [ "yes", "no" ]
```
could document `default: "no"`

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance_facts.py:
```diff
@@ -0,0 +1,150 @@
+# FIXME - return needs updated
+
+RETURN = '''
+instances:
```
`instances` -> `db_instances`

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance_facts.py:
```diff
@@ -0,0 +1,150 @@
+               "create_time": "2017-02-20T06:13:05.724000+00:00",
+               "db_engine": "postgres",
+               "endpoint": "helloworld-rds-master.cyaju3p4tyqv.ap-southeast-2.rds.amazonaws.com",
+               "id": "helloworld-rds-master",
```
id -> db_instance_identifier

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance_facts.py:
```diff
@@ -0,0 +1,150 @@
+    required: false
+    aliases:
+      - id
+  filters:
```
You could document that this is a dict.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance_facts.py:
```diff
@@ -0,0 +1,150 @@
+
+argument_spec = dict(
+    db_instance_identifier=dict(aliases=['id']),
+    filters=dict(type='list', default=[])
```
This should be type='dict', default={}

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/python
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
```
New copyright: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+    snapshot = get_snapshot(conn, snapshot_name)
+    if not snapshot:
+        try:
+            snapshot = conn.create_db_snapshot(DBSnapshotIdentifier=snapshot_name,
```
If there are tags they should be added to this call.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+    if not module.params.get('wait'):
+        return dict(changed=True)
+    try:
+        await_resource(conn, result, 'deleted', module)
```
`result` -> `snapshot`

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot.py:
```diff
@@ -0,0 +1,196 @@
+def await_resource(conn, resource_name, status, module):
+    wait_timeout = module.params.get('wait_timeout') + time.time()
+    resource = get_snapshot(conn, resource_name)
+    while wait_timeout > time.time() and resource['Status'] != status:
```
Since None['Status'] fails, need to check for resource first.

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_snapshot_facts.py:
```diff
@@ -0,0 +1,160 @@
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
```
New copyright: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright

mikedlr committed:
rds module - start merging review changes from ryansb et al

Commit:
Change MySQL engine case to lower case

RDS complains if you pass in 'MySQL' rather than 'mysql', and
also thinks you need to modify the DB engine when applying to
an existing DB.

Commit:
Use shared snake_dict_to_camel_dict

Don't reimplement the wheel, just adjust for the fully upper
case abbreviations

Commit:
Fix startswith typo

mikedlr committed:
rds tests - fix camel conversion testing to use current interface

mikedlr committed:
rds modules - fixes according to new sanity tests plus copyright update

mikedlr committed:
rds_instance - fix database options docs to match exactly options

mikedlr committed:
rds tests - reorder so variables are set inside test for stability

mikedlr committed:
rds modules and tests - cleanups from CI verification for 2.6

mikedlr committed:
rds modules and tests - further cleanups from CI verification for 2.6

mikedlr committed:
rds module utils - fix assertion syntax to modern

gundalow commented:
@mikedlr 
Thank you for your PR.
As part of reviewing the backlog of PRs we are looking at PRs older PRs that haven't been updated in a while

Given that:
* PR has needed a rebase since 2017
* This is a feature rather than a bug fix

Therefore I'm going to close this.

If you or anyone else wants to continue with this work then please do feel free to create a fresh PR and `@mention` the previous reviewers in here.

willthames commented:
@gundalow also all the work in this PR is merged in rds_instance and similar modules already :)
